src/toil/provisioners/aws/awsProvisioner.py

```diff
@@ -0,0 +1,325 @@
+                                          user_data=userData)
+        else:
+            requests = []
+            for attempt in retry_ec2(retry_while=expectedLaunchErrors):
+                with attempt:
+                    requests = ctx.ec2.request_spot_instances(price=spotBid, image_id=coreOSAMI,
+                                                              count=1, key_name=keyName,
+                                                              security_groups=[securityGroupName],
+                                                              instance_type=instanceType,
+                                                              instance_profile_arn=profileARN,
+                                                              user_data=userData,
+                                                              launch_group=securityGroupName)
+            assert requests
+            AWSProvisioner._waitOnRequests(ctx, requests)
+
+    @staticmethod
+    def destroyCluster(securityGroupName):
+        ctx = Context(availability_zone='us-west-2a', namespace='/')
+        instances = AWSProvisioner.__getNodesInCluster(ctx, securityGroupName, both=True)
+        spotIDs = AWSProvisioner._getSpotRequestIDs(ctx, securityGroupName)
+        if spotIDs:
+            ctx.ec2.cancel_spot_instance_requests(request_ids=spotIDs)
+        if instances:
+            AWSProvisioner._deleteIAMProfiles(instances=instances, ctx=ctx)
+            AWSProvisioner._terminateInstance(instances=instances, ctx=ctx)
+        logger.info('Deleting security group...')
+        for attempt in retry_ec2(retry_after=30, retry_for=300, retry_while=expectedShutdownErrors):
+            with attempt:
+                ctx.ec2.delete_security_group(name=securityGroupName)
+        logger.info('... Succesfully deleted security group')
+
+    @staticmethod
+    def _terminateInstance(instances, ctx):
+        instanceIDs = [x.id for x in instances]
+        logger.info('Terminating instance(s): %s', instanceIDs)
+        ctx.ec2.terminate_instances(instance_ids=instanceIDs)
+        logger.info('Instance(s) terminated.')
+
+    @staticmethod
+    def _deleteIAMProfiles(instances, ctx):
+        instanceProfiles = [x.instance_profile for x in instances]
+        for profile in instanceProfiles:
+            profile_name = profile['arn'].split('/', 1)[1]
+            try:
+                ctx.iam.remove_role_from_instance_profile(profile_name, 'toil-appliance-worker')
+            except BotoServerError as e:
+                if e.status == 404:
+                    ctx.iam.remove_role_from_instance_profile(profile_name, 'toil-appliance-leader')
+                else:
+                    raise
+            ctx.iam.delete_instance_profile(profile_name)
+
+    def _addNodes(self, instancesToLaunch, preemptable=None):
+        bdm = self._getBlockDeviceMapping()
+        profileID = str(time.time())
```

Comment:
This looks wrong. But I think we talked about this yesterday and you already realized that all workers should share an instance profile. I'm probably looking at stale code.